8264288: Performance issue with MethodHandle.asCollector

JornVernee opened the pull request:
This patch speeds up MethodHandle.asCollector handles where the array type is not Object[], as well as speeding up all collectors where the arity is greater than 10.

The old code is creating a collector handle by combining a set of hard coded methods for collecting arguments into an Object[], up to a maximum of ten elements, and then copying this intermediate array into a final array.

In principle, it shouldn't matter how slow (or fast) this handle is, because it should be replaced by existing bytecode intrinsification which does the right thing. But, through investigation it turns out that the intrinsification is only being applied in a very limited amount of cases: Object[] with max 10 elements only, only for the intermediate collector handles. Every other collector shape incurs overhead because it essentially ends up calling the ineffecient fallback handle.

Rather than sticking on a band aid (I tried this, but it turned out to be very tricky to untangle the existing code), the new code replaces the existing implementation with a collector handle implemented using a LambdaForm, which removes the need for intrinsification, and also greatly reduces code-complexity of the implementation. (I plan to expose this collector using a public API in the future as well, so users don't have to go through MHs::identity to make a collector).

The old code was also using a special lambda form transform for collecting arguments into an array. I believe this was done to take advantage of the existing-but-limited bytecode intrinsification, at least for Object[] with less than 10 elements.

The new code just uses the existing collect arguments transform with the newly added collector handle as filter, and this works just as well for the existing case, but as a bonus is also much simpler, since no separate transform is needed. Using the collect arguments transform should also improve sharing.

As a result of these changes a lot of code was unused and has been removed in this patch.

Testing: tier 1-3, benchmarking using TypedAsCollector (part of the patch), as well as another variant of the benchmark that used a declared static identity method instead of MHs::identity (not included). Before/after comparison of MethodHandleAs* benchmarks (no differences there).

Here are some numbers from the added benchmark:

Before:
```
Benchmark                                    Mode  Cnt    Score    Error  Units
TypedAsCollector.testIntCollect              avgt   30  189.156 �  1.796  ns/op
TypedAsCollector.testIntCollectHighArity     avgt   30  660.549 � 10.159  ns/op
TypedAsCollector.testObjectCollect           avgt   30    7.092 �  0.042  ns/op
TypedAsCollector.testObjectCollectHighArity  avgt   30   65.225 �  0.546  ns/op
TypedAsCollector.testStringCollect           avgt   30   28.511 �  0.243  ns/op
TypedAsCollector.testStringCollectHighArity  avgt   30   57.054 �  0.635  ns/op
```
(as you can see, just the Object[] with arity less than 10 case is fast here)
After:
```
Benchmark                                    Mode  Cnt  Score   Error  Units
TypedAsCollector.testIntCollect              avgt   30  6.569 � 0.131  ns/op
TypedAsCollector.testIntCollectHighArity     avgt   30  8.923 � 0.066  ns/op
TypedAsCollector.testObjectCollect           avgt   30  6.813 � 0.035  ns/op
TypedAsCollector.testObjectCollectHighArity  avgt   30  9.718 � 0.066  ns/op
TypedAsCollector.testStringCollect           avgt   30  6.737 � 0.016  ns/op
TypedAsCollector.testStringCollectHighArity  avgt   30  9.618 � 0.052  ns/op
```

Thanks,
Jorn

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264288](https://bugs.openjdk.java.net/browse/JDK-8264288): Performance issue with MethodHandle.asCollector


### Reviewers
 * [John R Rose](https://openjdk.java.net/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to 46bc5d199261d0f9cf590515b2a2eb4166be07d7
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3306/head:pull/3306` \
`$ git checkout pull/3306`

Update a local copy of the PR: \
`$ git checkout pull/3306` \
`$ git pull https://git.openjdk.java.net/jdk pull/3306/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3306`

View PR using the GUI difftool: \
`$ git pr show -t 3306`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3306.diff">https://git.openjdk.java.net/jdk/pull/3306.diff</a>

</details>

rose00 commented on src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:
```diff
@@ -648,57 +648,6 @@ LambdaForm collectArgumentsForm(int pos, MethodType collectorType) {
         return putInCache(key, form);
     }
 
-    LambdaForm collectArgumentArrayForm(int pos, MethodHandle arrayCollector) {
```
It's counter-intuitive that removing a LFE tactic would be harmless.
Each LFE tactic is a point where LFs can be shared, reducing footprint.
But in this case `collectArgumentArrayForm` is always paired with
`collectArgumentsForm`, so the latter takes care of sharing.
The actual code which makes the arrays is also shared, via
`Makers.TYPED_COLLECTORS` (unchanged).

rose00 commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -1527,34 +1327,25 @@ static MethodHandle varargsArray(Class<?> arrayType, int nargs) {
+        return true;
     }
 
+    // Array identity function (used as getConstantHandle(MH_arrayIdentity)).
```
If this is no longer used, remove it.

JornVernee replied:
Yes, it looks like this can be removed. Good catch!

rose00 commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -2089,21 +1880,73 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
+        return mh;
+    }
+
+    private static LambdaForm makeCollectorForm(MethodType basicType, MethodHandle storeFunc) {
```
This reads very well.  It is the right way to do this job with LFs.
There is a slight risk that the JIT will fail to inline the tiny MHs
that (a) construct arrays and (b) poke elements into them.
If that happens, then perhaps the bytecode generator can
treat them as intrinsics.  Maybe that's worth a comment in
the code, to help later readers?

JornVernee replied:
Good idea, I'll leave some notes.

PaulSandoz commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -1275,8 +1274,8 @@ String internalProperties() {
                 MethodType resultType = type().asCollectorType(arrayType, type().parameterCount() - 1, arrayLength);
-                MethodHandle newArray = MethodHandleImpl.varargsArray(arrayType, arrayLength);
-                return newArray.asType(resultType);
+                MethodHandle collector = MethodHandleImpl.makeCollector(arrayType, arrayLength);
```
Should `MethodHandleImpl.varargsArray` still be used here since it performs arity checking and caching? 

Maybe the arity checks are performed by `asCollectorType`, but that would still leave the caching.

JornVernee replied:
Ah right, good catch! This is a leftover from moving things around and should indeed use `varargsArray`.

rose00 reviewed (approved):
So many red deletion lines; I'm looking at a beautiful sunset!
It's a sunset for some very old code, some of the first code
I wrote for method handles, long before Lambda Forms
made this sort of task easier.

Thanks very much for cleaning this out.  See a few
minor comments on some diff lines.

PaulSandoz reviewed (commented):
That's an elegant solution.

At first i thought it might unduly perturb lambda form generation and caching. but you slotted a different lambda form implementation underneath the varargs implementation.

JornVernee commented:
I've address review comments, plus some other things:

- I realized that I was calling the uncached version of the store function factory. Fixed that.
- I also realized that there's already an `ARRAY_STORE` intrinsic, which I'm now using to avoid generating a call.
- I also realized that since we only have 1 array creation handle per lambda form, we can instead generate a direct call to `Array::newInstance` instead of going through the array constructor handle (which also avoids having to use a BoundMethodHandle).
- Finally, I added an instrinsic, under the old `NEW_ARRAY` name, that intrinsifies a call to `Array::newInstance` if the component type argument is constant (which it is in this case).

As a result, the lambda form is now fully intrinsified (no more calls in the generated bytecode) e.g.:

```
static java.lang.Object collector001_LLLL_L(java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object);
    Code:
       0: iconst_3
       1: anewarray     #12                 // class java/lang/String
       4: astore        4
       6: aload         4
       8: checkcast     #14                 // class "[Ljava/lang/String;"
      11: dup
      12: astore        4
      14: iconst_0
      15: aload_1
      16: checkcast     #12                 // class java/lang/String
      19: aastore
      20: aload         4
      22: iconst_1
      23: aload_2
      24: checkcast     #12                 // class java/lang/String
      27: aastore
      28: aload         4
      30: iconst_2
      31: aload_3
      32: checkcast     #12                 // class java/lang/String
      35: aastore
      36: aload         4
      38: areturn
```

Thanks,
Jorn

iwanowww commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -2089,21 +1876,64 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
+        // use the same underlying store function - by injecting an array creation handle as a
+        // field of the receiver handle.
+        // This is left as a followup enhancement, as it needs to be investigated if this causes
+        // profile pollution.
```
Profile pollution shouldn't be a problem here: both types (element type + array type) will be constants attached to the "receiver" BMH instance and during inlining will be fetched from there.  

I'm fine with handling it as a separate RFE.

JornVernee replied:
That's what I thought as well (but not 100% sure). I can partially roll back the last commit so the code still uses an injected array constructor handle, and then it should be easy to add caching in the cases where the component type is a reference type.

iwanowww replied:
Whatever you prefer. I'm fine with it either way.

JornVernee replied:
I found it easier to add sharing now, since it makes the intrinsic redundant (and it can be removed again).

JornVernee commented on src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:
```diff
@@ -870,12 +870,10 @@ void addMethod() {
                     continue;
                 case NEW_ARRAY:
-                    Class<?> rtype = name.function.methodType().returnType();
-                    if (isStaticallyNameable(rtype)) {
```
I removed the `NEW_ARRAY` intrinsic altogether at first, but added it back in the latest commit. I didn't add the check since I was not aware of the restriction.

Good to know, will add a check.

iwanowww replied:
Also, please, add a test case for such scenario. It should be trivial: just use a class defined by the test (loaded by `AppClassLoader`).

JornVernee commented:
Addressed latest review comments:
- Reverted back to using an injected constructor handle
- Added lambda form sharing
- Added a test case for a custom class to VarargsArrayTest, which catches the illegal access error in the intrinsified case (though the itrinsic itself is now removed).

I also had to switch back to the un-cached version for creating the array element setter, since the cached version casts the array to be a specific type, and if the lambda form is shared, this won't work (e.g. casting an Object[] to String[], depending on the order of creating the collectors). Adding caching there is left as a followup.

iwanowww commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -2089,21 +1875,90 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
+        // Other caching for primitive types happens at the MethodHandle level (see varargsArray).
+        boolean isSharedLambdaForm = parameterCount == 0 || basicType.parameterType(0) == Object.class;
+        if (isSharedLambdaForm) {
+            LambdaForm lform = basicType.form().cachedLambdaForm(MethodTypeForm.LF_COLLECTOR);
```
How does sharing work w.r.t. array store check? Unless you put `storeFunc` on the BMH, I don't see how sharing preserves proper check.

iwanowww replied:
Nevermind, it's the type of the method handle produced `makeCollector` which ensures that all the arguments have proper type.

iwanowww commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -2089,21 +1875,90 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
+        MethodType type = MethodType.methodType(arrayType, Collections.nCopies(parameterCount, arrayType.componentType()));
+
+        MethodHandle newArray = MethodHandles.arrayConstructor(arrayType);
+        MethodHandle storeFunc = ArrayAccessor.getAccessor(arrayType, ArrayAccess.SET);
```
I'd move `storeFunc` into `makeCollectorForm()` and use erased accessor there (`ArrayAccessor.OBJECT_ARRAY_SETTER`). 

(It is interesting that `OBJECT_ARRAY_SETTER` is wrapped into `Intrinsic.ARRAY_STORE` intrinsic.)

JornVernee replied:
Done. I think it's also more obvious this way that the erased setter is being used.

iwanowww reviewed (approved):
Looks good.

JornVernee commented:
Thanks for the reviews. I've submitted one more tier1-3 test run and if that's all green I'll go ahead and integrate this.

JornVernee commented:
/integrate
